src/test/util/blocktools.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2019 The Unit-e developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <test/util/blocktools.h>
+
+namespace blocktools {
+
+CBlockIndex *BlockIndexFake::MakeBlockIndex(const uint256 &hash, CBlockIndex *const prev) {
+  auto result = block_indexes.emplace(hash, CBlockIndex());
+  CBlockIndex *const index = &result.first->second;
+  index->phashBlock = &result.first->first;
+  index->nHeight = prev ? prev->nHeight + 1 : 0;
+  index->pprev = prev;
+  return index;
+}
+
+CBlockIndex *BlockIndexFake::Generate(const std::size_t count, const uint256 &starting_point) {
+  auto starting_block = block_indexes.find(starting_point);
+  std::size_t height = 0;
+  CBlockIndex *const starting_index = [&]() {
+    if (starting_block == block_indexes.end()) {
+      ++height;
+      return MakeBlockIndex(starting_point, nullptr);
+    } else {
+      return &starting_block->second;
+    }
+  }();
+  assert(starting_index);
+  assert(starting_index->phashBlock);
+
+  CBlockIndex *current_index = starting_index;
+  for (; height < count; ++height) {
+    const uint256 hash = GetRandHash();
+    current_index = MakeBlockIndex(hash, current_index);
+    assert(current_index);
+    assert(current_index->phashBlock);
+    assert(current_index->pprev);
+  }
+
+  return current_index;
+}
+
+std::shared_ptr<std::vector<CBlockIndex *>> BlockIndexFake::GetChain(const uint256 &tip_hash) {
+  auto tip = block_indexes.find(tip_hash);
+  auto result = std::make_shared<std::vector<CBlockIndex *>>();
+  if (tip == block_indexes.end()) {
+    return result;
+  }
+  CBlockIndex *current = &tip->second;
+  result->resize(current->nHeight + 1, nullptr);
+  std::size_t ix = current->nHeight;
+  do {
+    (*result)[ix] = current;
+    current = current->pprev;
+  } while (ix-- != 0);
+  assert(current == nullptr);
+  return result;
+}
+
+void BlockIndexFake::SetActiveChain(const CBlockIndex *tip, mocks::ActiveChainMock &active_chain_mock) {
+  assert(tip != nullptr);
+  std::shared_ptr<std::vector<CBlockIndex *>> active_chain = GetChain(tip->GetBlockHash());
+  active_chain_mock.mock_GetSize.SetStub([active_chain]() { return active_chain->size(); });
+  active_chain_mock.mock_GetHeight.SetStub([active_chain]() { return active_chain->size() - 1; });
+  active_chain_mock.mock_GetDepth.SetStub([active_chain](const blockchain::Height height) {
+    return active_chain->size() - height;
+  });
+  active_chain_mock.mock_AtHeight.SetStub([active_chain](const blockchain::Height height) -> const CBlockIndex * {
+    if (height >= active_chain->size()) {
+      return nullptr;
+    }
+    return active_chain->at(height);
+  });
+  active_chain_mock.mock_GetTip.SetStub([active_chain]() {
+    return active_chain->back();
```

Comment:
This and the next function will crash when called on empty active chain.

Reply:
Added a `BOOST_REQUIRE_MESSAGE` in 1c7a0463bbb927ab06c00035990d8cbbb17da045